californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -151,6 +153,7 @@ public void sendResponse(final Exchange exchange) {
 				// by a peer that wants to cancel the observation
 				// these NON notifications will later be removed from the
 				// exchange store when Exchange.setComplete() is called
+				addCleaner(exchange, response);
```

Comment:
If this NON notification fails to be sent, the server-side observe will be removed.

Reply:
I will add a test for this.

Reply:
I added the test without changing anything in my code and it seems to pass.
I will try to better understand this tomorrow.

Reply:
I rethink about that :
- the comment explain that the exchange is added only to be able to handle a RST to this notification. So if we don't succeed to send the notification there is no need to store an exchange to be able to handle a RST message.
- If my test  is correct,  it seems cleaning the exchange will not remove the observe relation.

So I think there is no issue here.

Reply:
:-).
Analysing it, only the exchange is completed (removing all old NON notifies), but the observe relation is currently not canceled on exchange completion.

Though some "resend responses scenarios" required a `Exchange.retransmitResponse()`, the new change is send and could then be rejected.

I changed to reject a previous notify, and then it fails.
(I also cleaned-up the test  little, see my PR #683, commit https://github.com/eclipse/californium/pull/683/commits/c29f1d0226061fc3b3ac4e8139fc072cbb945fb5)

Reply:
Just let me add:
Though the time  scope of exchanges and the effect of complete seems to have changed a lot over the time, I'm not sure, if the fact, that complete doesn't cancel the observe, is an bug or intended. AFAIK, the coap-server-side seems to keep the exchange incomplete by intention (see UdpMatcher.sendResponse()) and so I'm not sure, if it is a good idea to complete it on failure.

Reply:
> Though the time scope of exchanges and the effect of complete seems to have changed a lot over the time

This is not so clear to me but I believe before  there was the `complete()` and `completeCurrentRequest()` which meant 2 different things : complete the global/parent exchange or the child/inner one. In the current case I think we want to complete only the current request but I think we can do that anymore ? (completeCurrentRequest was removed right?)

Finally we always face the same issue about the exchange concept which hold to many concept ... This is an exchange at Message Coap sense, but also something which hold observe relation, but also transparent blockwise stuff. :'(
I really hope we can change this one day together.

> I changed to reject a previous notify, and then it fails.

That's right. So I can remove the `addCleaner` for notification (even for CON) but I suspect that there is a lot of exchanges which will not be clean as soon as it can be... but I suppose we could investigate this in another PR ...

Reply:
As mentioned in https://github.com/eclipse/californium/pull/675#issuecomment-402064904 with the "corner cases", I'm not sure, if moving the "clean-up" to the "add" pays-off.

It's not only the current implementation of the Exchange and complete.
That may be changed to reflect this awareness (e.g. call the `removeHandler` for the failed message). But make it a little more complex, what should happen, if sending a retransmission fails?
If you "complete" but afterwards receive a matching response?
E.g. the cloud server (usually the DTLS server) sends a request, by retransmission it reaches the "auto resumption" timeout and start a handshake, which fails. But the device received the request, reached also the "auto resumption" and then successfully answers the request, the server will not match that response. OK there may be many, many pitfalls, so I'm not sure, if this is realistic.
So FMPOV, just fixing the "aware leaks", as PR #683 ,has also it's pros.

Reply:
I think the issue you describe with retransmission will happen with both PR (this one and yours).

I don't even know what could be the behavior expected ? :tired_face: 
For example, we could also imagine that if a request failed to be send we don't want to stop the retransmission ... in that case "failedToSend" should not complete the exchange but be handle like a "sent" and so start timer for a retranmission ... but if there is an handhake with DTLS retransmission timeout, CoAP retransmission timing would not be not exactly what it is expected, They would be delayed by the DTLS retransmission timeout.

Reply:
> I think the issue you describe with retransmission will happen with both PR (this one and yours).

Sure, both don't handle such advanced and theoretical cases. 
What I want to express is, that the "move" doesn't really  pay off, because the rules for handling failures are not that simple.

Reply:
I get it but the californium code base is already very complex, and I still have the impression that we constantly add new corner case without succeed to make the code simpler. :worried: 

Anyway, if you think #683  is better, go for it.